pallets/maintenance-mode/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://docs.substrate.io/reference/frame-pallets/>
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+pub use weights::*;
+
+pub const ACTIVATED: bool = true;
+pub const DEACTIVATEDD: bool = false;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use cumulus_primitives_core::{
+		relay_chain::BlockNumber as RelayBlockNumber, DmpMessageHandler,
+	};
+	use frame_support::{
+		pallet_prelude::{ValueQuery, *},
+		traits::Contains,
+	};
+	use frame_system::pallet_prelude::*;
+	pub use log;
+	use sp_std::vec::Vec;
+	use xcm_primitives::PauseXcmExecution;
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[derive(Default)]
+	#[pallet::genesis_config]
+	/// Genesis config for maintenance mode pallet
+	pub struct GenesisConfig {}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			MaintenanceModeStatus::<T>::put(ACTIVATED);
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type MaintenanceModeOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		type FilteredCalls: Contains<Self::RuntimeCall>;
+		type MaintenanceDmpHandler: DmpMessageHandler;
+		type XcmExecutorManager: PauseXcmExecution;
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::storage]
+	pub type MaintenanceModeStatus<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		MaintenanceModeActivated,
+		MaintenanceModeDeactivated,
+		/// The call to suspend on_idle XCM execution failed with inner error
+		FailedToSuspendIdleXcmExecution {
+			error: DispatchError,
+		},
+		/// The call to resume on_idle XCM execution failed with inner error
+		FailedToResumeIdleXcmExecution {
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Maintenance mode was already activated
+		MaintenanceModeAlreadyActivated,
+		/// Maintenance mode was already deactivated
+		MaintenanceModeAlreadyDeactivated,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::activate_maintenance_mode())]
+		pub fn activate_maintenance_mode(origin: OriginFor<T>) -> DispatchResult {
+			T::MaintenanceModeOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				!MaintenanceModeStatus::<T>::get(),
+				Error::<T>::MaintenanceModeAlreadyActivated
+			);
+
+			MaintenanceModeStatus::<T>::put(ACTIVATED);
+
+			if let Err(error) = T::XcmExecutorManager::suspend_xcm_execution() {
```

Comment:
this error might be caused if the origin used is not allowed to resume or suspend the XCMP queue. I am not sure how much sense does it make to emit an event here, but I think it's worth putting it in a log as might indicate a configuration error.

Reply:
So the reason why the event is there is to inform the user that the maintenance mode was activated but it did not work for the DmpQueue. An error would revert the whole call (because the transactional macro) and in case of emergencies is better to have the extrinsic filtered. On the other hand, creating only a log might pass unperceived leaving the user to think that the maintenance was activated for both.